xarray/core/dataset.py

```diff
@@ -256,15 +257,23 @@ def merge_indexes(
                 levels.append(cat.categories)
 
             idx = pd.MultiIndex(levels, codes, names=names)
+            for n in names:
+                dims_to_replace[n] = dim
 
         vars_to_replace[dim] = IndexVariable(dim, idx)
         vars_to_remove.extend(var_names)
 
     new_variables = {k: v for k, v in variables.items() if k not in vars_to_remove}
     new_variables.update(vars_to_replace)
+
+    # update dimensions if necessary  GH: 3512
+    for k, v in new_variables.items():
+        if any(d in dims_to_replace for d in v.dims):
+            new_dims = [dims_to_replace.get(d, d) for d in v.dims]
+            new_variables[k] = type(v)(new_dims, v.data, attrs=v.attrs,
```

Comment:
FYI we could use `v._replace` here to save passing in the data that's not changing

Reply:
Thanks, @max-sixty 

Does `Variable` have `_replace` method? I don't see this in `variable.py`.
Yes, I wonder how can I smartly update just a dimension name.

Reply:
Ah, one second, let me make this!

Reply:
https://github.com/pydata/xarray/pull/3528